### ISSUE TITLE
Export Notes & Highlights

### ISSUE THREAD

**Issue by itprojects** [johnfactotum/foliate#93]
Please add a button in the Annotations button popup to export the notes to file.txt.
This way students, researchers, and even just enthusiast, could make even greater use of this already frighteningly productive app.

Proposed example inside file:

"Nothing can come of nothing."
Chapter A,
The Complete Works of William Shakespeare, 1616
William Shakespeare

"We know what we are, but know not what we may be.",
Chapter B,
The Complete Works of William Shakespeare, 1616
William Shakespeare

"Hell is empty and all the bugs are here."
Chapter C,
The Complete Works of William Shakespeare, 1616
William Shakespeare

**Comment by LoafingBunny** [comment]
Annotations are saved in a simple json format in .local/share/com.github.johnfactotum.Foliate/your-ebook-encoded-name.json

**Comment by itprojects** [comment]
The data from JSON file, it isn't directly usable.

One needs to manually parse the Note, Chapter, Book, Author, Year, Publisher etc.

All this has to be done before the book is deleted. Many might forget, or simply not know about it.

If this feature is part of the app - everything happens really quickly and easyly, saving lots of time later!

There is great value in such a feature.

**Comment by LoafingBunny** [comment]
Sure, didn't noticed text was missing from the json. Maybe would be safer to add annotated text too, in case the book is accidentally removed .
Too bad I'm not good with JavaScript and i cannot help

**Comment by itprojects** [comment]
The problems with accidentally removed books is discussed in https://github.com/johnfactotum/foliate/issues/73 , the discussion here is about a new feature: 'export notes and highlights'

**Comment by johnfactotum** [comment]
Texts are in the JSON file. See #11.

I agree it would be good to include book and chapter info, either in the file itself, or only when exported. The main reason why chapter info isn't already implemented is that there isn't currently a way to reliably get chapter names from CFIs. This is because sometimes a book would have multiple chapters within a spine item (books from Project Gutenberg are all like this).

For example, in [The Time Machine](https://www.gutenberg.org/ebooks/35), chapter two's href is `@public@vhost@g@gutenberg@html@files@35@35-h@35-h-0.htm.html#pgepubid00004`. Given a CFI within `[...]@35-h-0.htm.html`, there's currently no way of knowing whether the CFI is before or after the anchor, `#pgepubid00004`.

To correctly implement a "CFI to chapter name" function, one would mostly likely have to get the CFIs for all the anchors in the TOC. I'm not sure how to do that yet.

**Comment by itprojects** [comment]
To clarify: the idea is to export the information from the JSON, into a (non-JSON) citations file, without requiring the user to manually open the JSON and copy anything at all. Just press a button.

About https://github.com/johnfactotum/foliate/issues/11 "there isn't any standards for annotations" is not the case.

Take for example the now standardly used everywhere [Bibtex](https://www.verbosus.com/bibtex-style-examples.html). A very popular format for students and researchers that work with the TeX dialects, to produce pretty pdf's. 


@book{book,

  author    = {William Shakespeare}, 

  title     = {The title of the work},

  publisher = {The name of the publisher},

  year      = 1993,

  volume    = 4,

  series    = 10,

  address   = {The address},

  edition   = 3,

  month     = 7,

  note      = {An optional note},<-----------annotation +/- highlighted quote

  isbn      = {3257227892}

}


This will have to be done during the highlighting or note saving, otherwise it will take ages to parse slightly a longer list of notes.

Cross-referencing TOC and file refs is the right way to do it.

**Comment by LoafingBunny** [comment]
Partly agree, because what really matter when citing a book is the page number - all other information are valid for the entire book and can be gathered at any time. But is there a way to encode in a human readable way the page number of an epub?
Otherwise I just think it's a low priority issue - but of course anyone can send a pull request, implementation should be easy.
Maybe one of the next days I'll try myself

**Comment by itprojects** [comment]
Pages is the wrong way. Font sizes, image sizes, and scaling in general make it impossible to use pages. The extracts will never match.

Furthermore the chapter can be added to export later.
The initial feature should be export to external (plain or Bibtex) file.

**Comment by johnfactotum** [comment]
The reference to #11 is simply to clarify that the highlighted text is already saved in the JSON file. And by "no standard" I only meant there's no real standard when it comes to exporting annotations _to other ebook readers_ (in fact, there is an [Open Annotation](http://www.idpf.org/epub/oa/) standard, but as far as I know it isn't used by any ebook reader out there.), because that's what #11 was about, and it does not apply in the context of the current issue.

Whatever the format, it is indeed quite trivial to do exporting. The "page number" is saved as EPUB Canonical Fragment Identifiers (CFI). Any tool that supports CFIs will be able to process them and extract the necessary information to recreate the annotations in whatever format one desires. If all you need is the metadata plus the texts, then you don't even need support for CFI, because the texts are already available in the JSON file.

What is not that trivial about implementing this in Foliate itself is, as I have already mentioned, to get the correct chapter name, if that info is needed for whatever format the user is exporting to. I only mentioned this because it is what's included in @itprojects original "Proposed example inside file" above.

**Comment by itprojects** [comment]
Well, it's a proposed export file format, so we can just ignore Chapter for now.
This shouldn't affect the Bibtex.

For the CFI libraries - maybe we can use one (or it's code)? to extract the chapter...

**Comment by AnthonyBronislavaGonzales** [comment]
An epub reader with a nice annotation/note/bookmark/highlight exporter I've used is on Bluefire Reader for iOS. This is what it looks like:

![email-html export](https://funkyimg.com/i/2Vq1Y.jpg)

So this is clear, beautiful, and great for reading back through. I think they should also do something like this:

Highlight: _One day,_
Note: **Switched from the complete** 

etc. That would make searching easier.

This is their email export; they don't allow exporting as just an html file, but they should. Maybe something like an rtf export would be good too, but probably not necessary or called for in my opinion since html is enough for all the formatting needed. Unless there are other features people want?

They have a *.txt export too on the rudimentary windows version. This is what it looks like, by necessity they have the feature helpful for searching that I mentioned:

![text export example](https://funkyimg.com/i/2Vq24.jpg)

Not bad. The different sorts of notes are cleanly separated. Beyond annotations which pertain to highlighted text, you can make a note on a bookmark (location shown as the page number) as well. I know sometimes I would rather make a note general to a section or the place I'm at in the text, rather than on a highlight, but this isn't of dire importance, since you can just highlight a period or something if you want.

Other ideas: Moon Reader Pro+ on Android, in their export (which unfortunately uses parentheses to separate highlights from notes in their text export ... actual text often has parentheses, that's confusing. Don't use parentheses.)  --- give you a numeric count of how many Bookmarks, Highlights, and Notes you have in the document. I can see that being useful for people who want to comb through a bunch of notes, for example.

**Comment by itprojects** [comment]
The formatting you suggest looks very nice, and would be a great addition to the app, even with chapter marking missing initially.
